### PR TITLE
Support duration chaos for pod failure

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve? 
https://github.com/pingcap/chaos-mesh/issues/171
Make `Podfailure chaos` support duration chaos feature.

### What is changed and how does it work?
Adding a new controller interface as `duration controller` like `twophase controller` to control the logic of duration chaos. Each kind of chaos would implement both interfaces.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

Code changes

 - Has Go code change

Side effects

 - None

Related changes

 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
Support duration chaos for pod failure by omitting the scheduler and duration.
```

### CODE REVIEW

File: controllers/duration/podfailure/types.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package podfailure
+
+import (
+	"context"
+	"errors"
+	"github.com/go-logr/logr"
```
Comment (zhouqiang-cl) [inline controllers/duration/podfailure/types.go]: add a blank line?

**Comment by zhouqiang-cl** [comment]
@mahjonp @cwen0 @YangKeao PTAL

File: controllers/podchaos/podfailure/types.go [hunk 2]
```diff
@@ -15,7 +15,6 @@ package podfailure
 
 import (
```
Comment (zhouqiang-cl) [inline controllers/podchaos/podfailure/types.go]: should we make a dir named controllers/scheduler/podchaos

File: controllers/duration/types.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+	err = r.Apply(ctx, req, chaos)
+	if err != nil {
+		r.Log.Error(err, "failed to apply chaos action")
+
+		updateError := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			return r.Update(ctx, chaos)
+		})
+		if updateError != nil {
+			r.Log.Error(updateError, "unable to update chaos finalizers")
```
Comment (zhouqiang-cl) [inline controllers/duration/types.go]: Did we need return updateError here

File: controllers/podchaos/types.go [hunk 4]
```diff
@@ -40,7 +45,41 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+	case v1alpha1.PodKillAction:
+		return r.notSupportedResponse(podchaos), nil
+	case v1alpha1.PodFailureAction:
+		r := durationPodfailure.NewDurationReconciler(r.Client, r.Log.WithValues("action", "pod-failure"), req)
+		reconciler := duration.Reconciler{
+			InnerDurationReconcile: r,
+			Client:                 r.Client,
+			Log:                    r.Log,
+		}
+		return reconciler.Reconcile(req)
```
Comment (mahjonp) [inline controllers/podchaos/types.go]: I think this is a little redundant, how about merge these two parts into one constructor method?

File: controllers/podchaos/podfailure/types.go [hunk 5]
```diff
@@ -96,22 +86,17 @@ func (r *Reconciler) Apply(ctx context.Context, req ctrl.Request, chaos twophase
...
 
-func (r *Reconciler) Recover(ctx context.Context, req ctrl.Request, chaos twophase.InnerObject) error {
-	podchaos, ok := chaos.(*v1alpha1.PodChaos)
-	if !ok {
-		err := errors.New("chaos is not PodChaos")
-		r.Log.Error(err, "chaos is not PodChaos", "chaos", chaos)
-		return err
-	}
+// Clean would recover the podchaos for the selected pods
+func (r *Reconciler) Clean(ctx context.Context, req ctrl.Request, podchaos *v1alpha1.PodChaos) error {
```
Comment (mahjonp) [inline controllers/podchaos/podfailure/types.go]: I think it's is more suitable to put type assertion logic down to `Clean` and `Perform`. There are redundant codes in `duration/podfailure/types.go` and `twophase/podfailure/types.go`

**Comment by cwen0** [comment]
@Yisaer  I think the name `duration chaos` is not very accurate to describe this kind of chaos, because we don't have the option of duration to control this kind of chaos, which is easy to cause misunderstanding.

**Comment by cwen0** [comment]
@Yisaer I think we can rename `twophase controller` to `cheduler controller` and  rename `duration chaos` to `common chaos`. What do you think?

**Comment by Yisaer** [comment]
@cwen0 It's ok for me to rename both `twophase controller` and `duration controller`. In this request, I would rename `duration chaos` to `common chaos`. We could create new request for renaming `twophase controller`.

**Comment by cwen0** [comment]
@Yisaer Good!

**Comment by cwen0** [comment]
> @cwen0 It's ok for me to rename both `twophase controller` and `duration controller`. In this request, I would rename `duration chaos` to `common chaos`. We could create new request for renaming `twophase controller`.

Great！

File: controllers/duration/types.go (hunk 3 above)
Comment (Yisaer) [reply controllers/duration/types.go]: I think we don't need to, this reconciliation would be re queued in the end.

File: controllers/duration/podfailure/types.go (hunk 1 above)
Comment (Yisaer) [reply controllers/duration/podfailure/types.go]: updated.

File: controllers/podchaos/podfailure/types.go (hunk 5 above)
Comment (Yisaer) [reply controllers/podchaos/podfailure/types.go]: updated.

File: controllers/podchaos/podfailure/types.go [hunk 6]
```diff
@@ -96,20 +96,22 @@ func (r *Reconciler) Apply(ctx context.Context, req ctrl.Request, chaos twophase
...
-func (r *Reconciler) Recover(ctx context.Context, req ctrl.Request, chaos twophase.InnerObject) error {
-	podchaos, ok := chaos.(*v1alpha1.PodChaos)
+// Clean would recover the podchaos for the selected pods
+func (r *Reconciler) Clean(ctx context.Context, req ctrl.Request, obj apiinterface.StatefulObject) error {
+
+	podchaos, ok := obj.(*v1alpha1.PodChaos)
 	if !ok {
 		err := errors.New("chaos is not PodChaos")
-		r.Log.Error(err, "chaos is not PodChaos", "chaos", chaos)
+		r.Log.Error(err, "chaos is not PodChaos", "chaos", podchaos)
```
Comment (mahjonp) [inline controllers/podchaos/podfailure/types.go]:
```suggestion
		r.Log.Error(err, "chaos is not PodChaos", "chaos", obj)
```

File: controllers/podchaos/podfailure/types.go [hunk 7]
```diff
@@ -60,15 +58,18 @@ type Reconciler struct {
...
-func (r *Reconciler) Apply(ctx context.Context, req ctrl.Request, chaos twophase.InnerObject) error {
-	podchaos, ok := chaos.(*v1alpha1.PodChaos)
+// Perform would perform the podfailure chaos for the selected pods
+func (r *Reconciler) Perform(ctx context.Context, req ctrl.Request, obj apiinterface.StatefulObject) error {
+
+	podchaos, ok := obj.(*v1alpha1.PodChaos)
 	if !ok {
 		err := errors.New("chaos is not PodChaos")
-		r.Log.Error(err, "chaos is not PodChaos", "chaos", chaos)
+		r.Log.Error(err, "chaos is not PodChaos", "chaos", podchaos)
```
Comment (mahjonp) [inline controllers/podchaos/podfailure/types.go]:
```suggestion
		r.Log.Error(err, "chaos is not PodChaos", "chaos", obj)
```

File: controllers/podchaos/types.go [hunk 8]
```diff
@@ -40,7 +46,37 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		r.Log.Error(err, "unable to get podchaos")
 		return ctrl.Result{}, nil
 	}
+	scheduler := podchaos.GetScheduler()
+	duration, err := podchaos.GetDuration()
+	if err != nil {
+		r.Log.Error(err, "unable to get duration")
```
Comment (mahjonp) [inline controllers/podchaos/types.go]: carry detailed err?

File: controllers/podchaos/types.go [hunk 9]
```diff
@@ -40,7 +46,37 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+		return ctrl.Result{}, nil
+	}
+	if scheduler == nil && duration == nil {
+		return r.durationPodChaos(&podchaos, req)
```
Comment (mahjonp) [inline controllers/podchaos/types.go]: maybe you should give it a better name than `durationPodChaos`?

**Comment by you06** [comment]
I find both `twophase controller` and `duration controller` wrap `podfailure` in very similar ways and they do only expose some functions from `podfailure`, what's the purpose for it?

File: controllers/podchaos/types.go (hunk 9 above)
Comment (Yisaer) [reply controllers/podchaos/types.go]: updated.

File: controllers/podchaos/types.go (hunk 8 above)
Comment (Yisaer) [reply controllers/podchaos/types.go]: updated.

**Comment by Yisaer** [comment]
> I find both `twophase controller` and `duration controller` wrap `podfailure` in very similar ways and they do only expose some functions from `podfailure`, what's the purpose for it?

The wrap here is to implement the interface like `InnerReconciler` and `InnerCommonReconcile` which were used to handle the logic of `common chaos` and `scheduler chaos`.

File: controllers/podchaos/types.go [hunk 10]
```diff
@@ -40,7 +46,37 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+		return ctrl.Result{}, nil
+	}
+	if scheduler == nil && duration == nil {
+		return r.commonPodChaos(&podchaos, req)
+	} else if scheduler != nil && duration != nil {
+		return r.schedulePodChaos(&podchaos, req)
+	}
 
+	// This should be ensured by admission webhook in the future
+	r.Log.Error(fmt.Errorf("podchaos[%s/%s] spec invaild", podchaos.Namespace, podchaos.Name), "scheduler and duration should be omiited or defined at the same time")
```
Comment (mahjonp) [inline controllers/podchaos/types.go]:
```suggestion
	r.Log.Error(fmt.Errorf("podchaos[%s/%s] spec invalid", podchaos.Namespace, podchaos.Name), "scheduler and duration should be omitted or defined at the same time")
```

**Comment by zhouqiang-cl** [comment]
@cwen0 PTAL again

File: controllers/common/podfailure/types.go [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+package podfailure
+
+import (
+	"context"
+	"github.com/go-logr/logr"
+	"github.com/pingcap/chaos-mesh/controllers/common"
+	"github.com/pingcap/chaos-mesh/controllers/podchaos/podfailure"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
Comment (cwen0) [inline controllers/common/podfailure/types.go]: format imports 
```
	"context"

	"github.com/go-logr/logr"
	"github.com/pingcap/chaos-mesh/controllers/common"
	"github.com/pingcap/chaos-mesh/controllers/podchaos/podfailure"

	ctrl "sigs.k8s.io/controller-runtime"
	"sigs.k8s.io/controller-runtime/pkg/client"
```

File: controllers/common/types.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
```
Comment (cwen0) [inline controllers/common/types.go]: format imports

File: controllers/podchaos/podfailure/types.go [hunk 13]
```diff
@@ -19,12 +19,13 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/pingcap/chaos-mesh/pkg/apiinterface"
```
Comment (cwen0) [inline controllers/podchaos/podfailure/types.go]: format

File: controllers/podchaos/podfailure/types.go [hunk 14]
```diff
@@ -60,15 +58,18 @@ type Reconciler struct {
 	Log logr.Logger
 }
 
-func (r *Reconciler) Object() twophase.InnerObject {
+// Instance return the instance of PodChaos
```
Comment (cwen0) [inline controllers/podchaos/podfailure/types.go]: s/return/returns/g

File: controllers/podchaos/types.go [hunk 15]
```diff
@@ -40,7 +46,37 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		r.Log.Error(err, "unable to get podchaos")
 		return ctrl.Result{}, nil
 	}
+	scheduler := podchaos.GetScheduler()
+	duration, err := podchaos.GetDuration()
+	if err != nil {
+		r.Log.Error(err, fmt.Sprintf("unable to get podchaos[%s/%s]'s duration", podchaos.Namespace, podchaos.Name))
+		return ctrl.Result{}, nil
+	}
+	if scheduler == nil && duration == nil {
```
Comment (cwen0) [inline controllers/podchaos/types.go]: How about `pod-kill` action? `Duration` is always empty for `pod-kill`

File: controllers/twophase/podfailure/types.go [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package podfailure
+
+import (
+	"context"
```
Comment (cwen0) [inline controllers/twophase/podfailure/types.go]: format it

File: controllers/podchaos/types.go (hunk 15 above)
Comment (Yisaer) [reply controllers/podchaos/types.go]: good catch and updated.

File: controllers/podchaos/podfailure/types.go (hunk 14 above)
Comment (Yisaer) [reply controllers/podchaos/podfailure/types.go]: updated.

File: controllers/common/podfailure/types.go (hunk 11 above)
Comment (Yisaer) [reply controllers/common/podfailure/types.go]: updated.

File: controllers/twophase/podfailure/types.go (hunk 16 above)
Comment (Yisaer) [reply controllers/twophase/podfailure/types.go]: updated.

File: controllers/common/types.go (hunk 12 above)
Comment (Yisaer) [reply controllers/common/types.go]: updated.

File: controllers/podchaos/podfailure/types.go (hunk 13 above)
Comment (Yisaer) [reply controllers/podchaos/podfailure/types.go]: updated.

**Comment by zhouqiang-cl** [comment]
@cwen0 PTAL

**Pull request merged**